agent/pkg/api/manage_agent.go

```diff
@@ -0,0 +1,50 @@
+package api
+
+import (
+	"context"
+	"io"
+	"time"
+
+	pb "github.com/buoyantio/linkerd-buoyant/gen/bcloud"
+)
+
+func (c *Client) RecvCommand() (*pb.AgentCommand, error) {
+	for {
+		if c.manageAgentClient == nil {
+			c.manageAgentClient = c.newManageAgentClient()
+		}
+
+		msg, err := c.manageAgentClient.Recv()
+		if err == io.EOF {
+			c.log.Info("ManageAgentStream closed")
+			c.manageAgentClient = nil
+			continue
+		} else if err != nil {
+			c.log.Errorf("ManageAgentClient failed to receive: %s", err)
+			return nil, err
+		}
+		return msg, nil
+	}
+}
+
+func (c *Client) newManageAgentClient() pb.Api_ManageAgentClient {
+	var client pb.Api_ManageAgentClient
+
+	// loop until the request to initialize a client succeeds
+	for {
+		var err error
+		client, err = c.client.ManageAgent(context.Background(), c.auth)
+		if err != nil {
+			c.log.Errorf("failed to initialize client: %s", err)
+			time.Sleep(100 * time.Millisecond)
+			continue
+		}
+
+		c.log.Info("ManageAgentClient initialized")
+
+		break
+	}
+
+	c.log.Info("ManageAgentClient connected")
+	return client
+}
```

Comment:
I agree we can simplify things a bit here. However, I am not sure we actually need a mutex. I believe that the workloads stream needs to be guarded by a mutex because it is being accessed by callbacks coming from the shared informers machinery which runs in separate go routines. So there may be concurrent calls to `sendMessage` and  `resetStream`. With the agent commands stream though we are just pulling messages from one go routine so I do not see much point in using any synchronization here. Am I missing something here?

Reply:
I actually like the older approach quite a bit. At least all the logic is isolated in one place. I think I will adopt that for both streams